packages/firestore/src/local/indexeddb_schema.ts

```diff
@@ -432,40 +454,144 @@ export class DbTargetGlobal {
-     * targets resumed with a resumeToken) should be suppressed (buffered)
-     * until the backend has caught up to this snapshot version again. This
-     * prevents our cache from ever going backwards in time.
+     * @param lastRemoteSnapshotVersion - A global snapshot version representing
+     * the last consistent snapshot we received from the backend. This is
+     * monotonically increasing and any snapshots received from the backend
+     * prior to this version (e.g. for targets resumed with a resumeToken)
+     * should be suppressed (buffered) until the backend has caught up to this
+     * snapshot version again. This prevents our cache from ever going backwards
+     * in time.
      */
     public lastRemoteSnapshotVersion: DbTimestamp
   ) {}
 }
 
+function createQueryCache(db: IDBDatabase): void {
+  const targetDocumentsStore = db.createObjectStore(DbTargetDocument.store, {
+    keyPath: DbTargetDocument.keyPath as KeyPath
+  });
+  targetDocumentsStore.createIndex(
+    DbTargetDocument.documentTargetsIndex,
+    DbTargetDocument.documentTargetsKeyPath,
+    { unique: true }
+  );
+
+  const targetStore = db.createObjectStore(DbTarget.store, {
+    keyPath: DbTarget.keyPath
+  });
+
+  // NOTE: This is unique only because the TargetId is the suffix.
+  targetStore.createIndex(
+    DbTarget.queryTargetsIndexName,
+    DbTarget.queryTargetsKeyPath,
+    { unique: true }
+  );
+  db.createObjectStore(DbTargetGlobal.store);
+}
+
 /**
- * The list of all IndexedDB stored used by the SDK. This is used when creating
- * transactions so that access across all stores is done atomically.
+ * An object representing the changes at a particular snapshot version for the
+ * given target. This is used to facilitate storing query changelogs in the
+ * targetChanges object store.
+ *
+ * PORTING NOTE: This is used for change propagation during multi-tab syncing
+ * and not needed on iOS and Android.
  */
-export const ALL_STORES = [
+export class DbTargetChange {
+  /** Name of the IndexedDb object store.  */
+  static store = 'targetChanges';
+
+  /** Keys are automatically assigned via the targetId and snapshotVersion. */
+  static keyPath = ['targetId', 'snapshotVersion'];
+
+  constructor(
+    /**
+     * @param targetId - The targetId identifying a target.
+     */
+    public targetId: TargetId,
+    /**
+     * @param snapshotVersion - The snapshot version for this change.
+     */
+    public snapshotVersion: DbTimestamp,
+    /**
+     * @param changes - The keys of the changed documents in this snapshot.
+     */
+    public changes: {
+      added?: EncodedResourcePath[];
+      modified?: EncodedResourcePath[];
+      removed?: EncodedResourcePath[];
+    }
+  ) {}
+}
+
+function createTargetChangeStore(db: IDBDatabase): void {
+  db.createObjectStore(DbTargetChange.store, {
+    keyPath: DbTargetChange.keyPath as KeyPath
+  });
+}
+
+/**
+ * A record of the metadata state of each instance.
```

Comment:
I think 'instance' is a pretty ambiguous term and so it would be good to at least occasionally in comments clarify, e.g. "of each instance of the Firestore client (i.e. running in different browser tabs)."

Reply:
Changed the naming here and elsewhere from `instance` to `client` as per our discussion.